Reframe in terms of groups ++ (#22)
cfrg/draft-irtf-cfrg-hybrid-kems: draft-irtf-cfrg-hybrid-kems.md

# Introduction {#intro}

> Post-quantum (PQ) cryptographic schemes offer new constructions based on problems conjectured as resistant to attacks possible on a quantum computer. Key Encapsulation Mechanisms
> (KEMs), are a standardized algorithm type that can be used to build protocols in

Reviewer: ```suggestion
Post-quantum (PQ) cryptographic schemes offer new constructions based on problems
conjectured as resistant to attacks possible on a quantum computer. Key Encapsulation
Mechanisms (KEMs), are a standardized class of cryptographic scheme that can be used
to build protocols in
```

# Introduction {#intro}

Post-quantum (PQ) cryptographic schemes offer new constructions based on problems conjectured as resistant to attacks possible on a quantum computer. Key Encapsulation Mechanisms
(KEMs), are a standardized algorithm type that can be used to build protocols in
lieu of traditional, quantum-vulnerable variants such as finite field or

> elliptic curve Diffie-Hellman (DH) based protocols. Upgrading key establishment
> protocols to use PQ KEMs is a priority for the protocol design community, due to
> the possibility of "harvest now, decrypt later" attacks.

Reviewer: ```suggestion
lieu of traditional, quantum-vulnerable variants such as finite field or
elliptic curve Diffie-Hellman (DH) based protocols.
```

# Introduction {#intro}

Post-quantum (PQ) algorithms offer new constructions based on problems tailored
towards resisting attack from a quantum computer. Key Encapsulation Mechanisms
(KEMs), are a standardized algorithm type that can be used to build protocols in
lieu of traditional, quantum-vulnerable variants such as finite field or
elliptic curve Diffie-Hellman (DH) based protocols. Upgrading key establishment
protocols to use PQ KEMs is a priority for the protocol design community, due to

> the possibility of "harvest now, decrypt later" attacks.

Reviewer 2: Recommend removal of HNDL attacks. Motivation for PQC is not specific to that, but due to quantum threats.

Author: The idea of mentioning HNDL is just to explain why the focus on KEMs vs. Authentication.  But I don't feel super strongly.

Author: Fine, fine, we can delete it :)

> Post-quantum (PQ) cryptographic schemes offer new constructions based on problems
> conjectured as resistant to attacks possible on a quantum computer. Key Encapsulation
> Mechanisms (KEMs), are a standardized class of cryptographic scheme that can be used
> to build protocols in
> lieu of traditional, quantum-vulnerable variants such as finite field or
> lieu of traditional, quantum-vulnerable variants such as finite field or
> elliptic curve Diffie-Hellman (DH) based protocols.

Reviewer: ```suggestion
Post-quantum (PQ) cryptographic schemes offer new constructions based on problems
conjectured as resistant to attacks possible on a quantum computer. Key 
Encapsulation Mechanisms (KEMs), are a standardized class of cryptographic scheme
that can be used to build protocols in lieu of traditional, quantum-vulnerable
variants such as finite field or elliptic curve Diffie-Hellman (DH) based protocols.
```

this risk while still providing security against quantum attack.  If construted
properly, a hybrid KEM will retain certain security properties even if one of
the two constituent KEMs is compromised.  If the PQ KEM is broken, then the
hybrid KEM should continue to provide security against non-quantum attackers by
virtue of its traditional KEM component.  If the traditional KEM is broken by a
quantum computer, then the hybrid KEM should continue to resist quantum attack
by virtue of its PQ KEM component.

In addition to guarding against algorithm weaknesses, this property also guards
against flaws in implementations, such as timing attacks.  Hybrid KEMs can also
facilitate faster deployment of PQ security by allowing applications to
incorporate PQ algorithms while still meeting compliance requirements based on
traditional algorithms.

In this document, we define generic schemes for constructing hybrid KEMs from a
traditional algorithm and a PQ KEM.  The aim of this document is provide a small
set of techniques to achieve specific security properties given conforming
component algorithms, which should make these techniques suitable for a broad
variety of use cases.

> The remainder of this document is structured as follows: First, in
> {{cryptographic-deps}} and {{schemes}}, we define the abstractions on which the
> schemes are built, and then the schemes themselves.  Then, in {{security}}, we

Reviewer: ```suggestion
The remainder of this document is structured as follows: first, in
{{cryptographic-deps}} and {{schemes}}, we define the abstractions on which the
schemes are built, and then the schemes themselves.  Then, in {{security}}, we
```

  by a cryptographically-secure random number generator.
- `concat(x0, ..., xN)`: Concatenation of byte strings.  `concat(0x01,
  0x0203, 0x040506) = 0x010203040506`.
- `split(N1, N2, x)`: Split a byte string `x` of length `N1 + N2` into its first
  `N1` bytes and its last `N2` bytes.  This function is the inverse of
  `concat(x1, x2)` when `x1` is `N1` bytes long and `x2` is `N2` bytes long. It
  is an error to call this function with a byte string that does not have length

>   `N1 + N2`.

Reviewer: So this is tricky, it's being used to split secret key material and public key material, and being leaky on one is acceptable and being leaky on the other is not...

# Notation

This document is consistent with all terminology defined in
{{?I-D.ietf-pquip-pqt-hybrid-terminology}}.

The following terms are used throughout this document:

- `random(n)`: return a pseudorandom byte string of length `n` bytes produced
  by a cryptographically-secure random number generator.
- `concat(x0, ..., xN)`: Concatenation of byte strings.  `concat(0x01,
  0x0203, 0x040506) = 0x010203040506`.
- `split(N1, N2, x)`: Split a byte string `x` of length `N1 + N2` into its first
  `N1` bytes and its last `N2` bytes.  This function is the inverse of
  `concat(x1, x2)` when `x1` is `N1` bytes long and `x2` is `N2` bytes long. It
  is an error to call this function with a byte string that does not have length
  `N1 + N2`. Since this function operates over secret data `x`, it MUST be

>   constant-time for a given `N1` and `N2`.

Reviewer: can `split()` be `split(x, N1, N2)`? Why are the input byte to be split last?

The generic hybrid PQ/T KEM constructions we define depend on the the
following cryptographic primitives:

> - Key Encapsulation Mechanisms {{kems}}
> - Nominal Groups {{group}}
> - Hash Functions {{hash}}

Reviewer: ```suggestion
- Key Encapsulation Mechanisms {{kems}}
- Nominal Groups {{group}}
- Key-derivation Functions {{kdfs}}
- Extendable-output Functions {{xofs}}
- Hash Functions {{hash}}
```

The generic hybrid PQ/T KEM constructions we define depend on the the
following cryptographic primitives:

- Key Encapsulation Mechanisms ({{kems}})
- Nominal Groups ({{groups}})

> - Key Derivation Functions ({{kdfs}})

Reviewer: <3

> - Key Derivation Functions ({{kdfs}})

Reviewer: ```suggestion
- Key Derivation Functions ({{kdfs}})
- Extendable-Output Functions ({{xofs}})
```

   `ek` and randomness `randomness`, and outputs a ciphertext `ct` and shared
   secret `shared_secret`.

<!-- XXX(RLB): Maybe we should make this optional and parallel to `EncapsDerand`
-->

We assume that the values produced and consumed by the above functions are all
byte strings, with fixed lengths:

- `Nseed`: The length in bytes of a key seed (input to DeriveKeyPair)
- `Nek`: The length in bytes of a public encapsulation key
- `Ndk`: The length in bytes of a secret decapsulation key
- `Nct`: The length in bytes of a ciphertext produced by Encaps
- `Nss`: The length in bytes of a shared secret produced by Encaps or Decaps

> This interface is effectively the same as the one defined in the Hybrid Public
> Key Encryption (HPKE) specification {{?RFC9180}}.  The only difference is that
> here we assume that all values are byte strings, whereas in HPKE keys are
> opaque by default and serialized or deserialized as needed.  We also use
> slightly different terminology for keys, emphasizing "encapsulation" and

> > "decapsulation" as opposed to "public" and "secret".
> 
> ## Hash functions {#hash}

Reviewer: ```suggestion
> "decapsulation" as opposed to "public" and "secret".

## Key Derivation Function `KDF` {#kdf}

A secure key derivation function (KDF) that is modeled as a secure
pseudorandom function (PRF) in the standard model {{GHP2018}} and
independent random oracle in the random oracle model (ROM) and quantum
random oracle model. Generally a strong KDF will have a proof of
indifferentiability from a random oracle.

Examples of secure KDFs in practice include HKDF-SHA256 and SHA3.
SHA-256 is not generally considered a strong KDF except under
constrained circumstances {{CDMP2005}}.

## Extendable-output function `XOF` {#xof}

Extendable-output function (XOF). A function on bit strings in which the
output can be extended to any desired length. Ought to satisfy the following
properties as long as the specified output length is sufficiently long to
prevent trivial attacks:

1. (One-way) It is computationally infeasible to find any input that maps to
   any new pre-specified output.

2. (Collision-resistant) It is computationally infeasible to find any two
   distinct inputs that map to the same output.

MUST provide the bit-security required to source input randomness for PQ/T
components from a seed that is expanded to a output length, of which a subset
is passed to the component key generation algorithms.

## Hash functions {#hash}
```

Reviewer: @br-hale @bwesterb thoughts?

* HashEverything - A generic construction that is suitable for use with any choice
  of traditional and PQ KEMs, with minimal security assumptions on the
  constituent KEMs
* PreHashedKeys - A performance optimization of HashEverything for the case
  where encapsulation keys are large and frequently reused
* HashTraditionalOnly - An optimized generic construction for the case where the traditional
  component is a nominal group and the PQ component has strong binding
  properties

In this section, we define a collection of constructions for hybrid KEMs. These
constructions share a common overall structure, differing mainly in how they
compute the final shared secret.

## HashEverything

The HashEverything hybrid KEM depends on the following constituent components:

* `KEM_T`: A traditional KEM
* `KEM_PQ`: A post-quantum KEM
* `ExpandHash`: A hash function mapping byte strings of length `KEM_H.Nseed` to
  byte strings of length `KEM_T.Nseed + KEM_PQ.Nseed` (`ExpandHash.Nh ==
  KEM_T.Nseed + KEM_PQ.Nseed`)
* `CombineHash`: A hash function mapping byte strings of length `KEM_T.Nss +
  KEM_PQ.Nss` to byte strings of length `KEM_H.Nss` (`CombineHash.Nh ==

>   KEM_H.Nss`)

Reviewer 3: @dconnolly citation?

  constituents being used.

We presume the KEMs and hash functions meet the interfaces described in
{{cryptographic-deps}}.

The constants associated with the hybrid KEM are mostly derived from the
concatenation of keys and ciphertexts:

~~~
Nek = KEM_T.Nek + KEM_PQ.Nek
Ndk = KEM_T.Ndk + KEM_PQ.Ndk
Nct = KEM_T.Nct + KEM_PQ.Nct
~~~

The `Nseed` and `Nss` constants should reflect the overall security level of the
combined KEM, with the following recommended values:

~~~
Nseed = max(KEM_T.Nseed, KEM_PQ.Nseed)
Nss = min(KEM_T.Nss, KEM_PQ.Nss)
~~~

Given these constituent parts, we define the HashEverything hybrid KEM as
follows:

~~~
def GenerateKeyPair():
    seed = random(Nseed)
    return DeriveKeyPair(seed)

def DeriveKeyPair(seed):
    seed_full = ExpandHash(seed)
    (seed_T, seed_PQ) = split(KEM_T.Nseed, KEM_PQ.Nseed, seed)
    (ek_T, dk_T) = KEM_T.DeriveKeyPair(seed_T)
    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
    ek_H = concat(ek_T, ek_PQ)
    dk_H = concat(dk_T, dk_PQ)
    return (ek_H, dk_H)

def Encaps(ek):
    (ek_T, ek_PQ) = split(KEM_T.Nek, KEM_PQ.Nek, ek)
    (ss_T, ct_T) = KEM_T.Encap(pk_T)
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(pk_PQ)

>     ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))

Reviewer: Where is `CombinerHash` defined?

    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):

>     (dk_T, dk_PQ) = split(KEM_T.Ndk, KEM_PQ.Ndk, dk)

Reviewer 3: +1 to @bifurcation -- since these are being thrown into the hash, parsing them at this layer seems necessary.

>     (dk_T, dk_PQ) = split(KEM_T.Ndk, KEM_PQ.Ndk, dk)

Reviewer: fair enough

    ss_T = KEM_T.Decap(dk_T, ct_T)
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))
    return ss_H
~~~

### Optimization for Encapsulation Key Reuse

The PreHashedKeys hybrid KEM is a performance optimization of the HashEverything
KEM, optimized for the case where encapsulation keys are large and frequently
reused. In such cases, hashing the entire encapsulation key is expensive, and
the same value is hashed repeatedly.  The PreHashedKeys KEM thus computes an
intermediate hash of the hybrid encapsulation key, so that the hash value can be
computed once and used across many encapsulation or decapsulation operations.

The PreHashedKeys KEM is identical to the HashEverything KEM except for the
shared secret computation.  One additional hash function is required:

* `KeyHash`: A hash function mapping byte strings of length `KEM_T.Nek +
  KEM_PQ.Nek` to byte strings of length `KEM_H.Nss` (`CombineHash.Nh ==
  KEM_H.Nss`)

The shared secret computation in `Encaps` and `Decaps` is then modified as
follows:

~~~
ekh = KeyHash(concat(ek_T, ek_PQ))
ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))
~~~

> ### Instantiation with a Nominal Group
> 
> The HashEverything and PreHashedKeys hybrid KEMs can be instantiated with a
> nominal group in place of the traditional KEM.  However, because the group
> construction used as a constituent does not comprise a secure KEM, these
> constructions are defined and analyzed separately.

Reviewer: ```suggestion
```

To instantiate the HashEverything or PreHashedKeys hybrid KEM with a nominal
group, the traditional KEM `KEM_T` is replaced with a "pseudo-KEM" `PseudoKEM_T`
based on a nominal group `Group_T`.  We use the term "pseudo-KEM" because while
this construction meets the KEM API definition, it is not a secure KEM.

> The constants for the pseudo-KEM are defined as follows:

Reviewer: ```suggestion
```

~~~
PseudoKEM_T.Nseed = Group_T.Nseed
PseudoKEM_T.Nek = GroupT.Nelem
PseudoKEM_T.Ndk = GroupT.Nscalar
PseudoKEM_T.Nct = GroupT.Nelem
PseudoKEM_T.Nss = GroupT.Nss
~~~

The pseudo-KEM interface methods are defined as follows:

~~~
def PseudoGenerateKeyPair():
    return DeriveKeyPair(random(Group_T.Nseed))

def PseudoDeriveKeyPair(seed):
    dk = Group_T.RandomScalar(seed)
    ek = Group_T.Exp(Group_T.g, dk)
    return (ek, dk)

def PseudoEncaps(ek):
    ct, sk_E = GenerateKeyPair()
    ss = GroupT.ElementToSharedSecret(Group_T.Exp(ek, sk_E))
    return (ct, ss)

def PseudoDecaps(dk, ct):
    return GroupT.ElementToSharedSecret(Group_T.Exp(ct, dk))

> ~~~

Reviewer: ```suggestion
```

> The HashEverything and PreHashedKeys hybrid KEMs are instantiated with a
> nominal group by replacing the `KEM_T` methods with the corresponding `Pseudo`
> methods.

Reviewer: ```suggestion
```

> ## HashTraditionalOnly

Reviewer: I don't love this name (I don't quite understand what it's trying to say) and think `QSF` is as good a name as any, short, and unless you already know it's associated with X-Wing, non-controversial?

Author: @BTW, what it's trying to say is that the only metadata you hash in is on the T side.  🤷

Reviewer 3: Let's leave it as-is and bike shed the name later. I think it's clear enough for now.

Reviewer: Yes leaving it the existing document names as-is seems fine (QSF)

> here we assume that all values are byte strings, whereas in HPKE keys are
> opaque by default and serialized or deserialized as needed.  We also use
> slightly different terminology for keys, emphasizing "encapsulation" and
> "decapsulation" as opposed to "public" and "secret".

## Hash functions {#hash}

Functionally, a hash function is simply a function that produces a fixed-length
output byte string from an input byte string of arbitrary length.

- `Nh` - The length in bytes of an output from this hash function.
- `Hash(input) -> output`: Produce a byte string of length `Nh` from an input
  byte string.

For simplicity, we will write invocations of a hash function without `Hash`
being explicit.  Invoking a hash function `Foo` on input `input` will be written
as `Foo(input)` instead of `Foo.Hash(input)`.

Hash function used with the constructions in this document should be structured
so that they can be regarded as random oracles in either the classical or
quantum random oracle model.  (Note that this property implies standard hash
function properties such as collision resistance and second preimage
resistance.)  Each hash function we refer to should be an independent random

> oracle.

Reviewer 2: Yes, there is a very significant difference for security. The descriptions (generic and concrete) need to match the generic building blocks used in the analyses. 

ALSO, we need to look at and discuss if the write concrete choices are used for the end goal (e.g., if something is using SHA256, SHA384, or SHA3). This may or may not be the same as used in the analysis - which means that if we change a building block from what is used in the analysis we need to take a closer look that the right properties are statisfied. 

I am also concerned the properties stated (e.g., collision resistance and second preimage resis) may not align to the properties needed for security, i.e., per what analysis has shown. We need to cite the write attributes.

> oracle.

Reviewer: Agreed; 'KitchenSink'/'HashEverything' seems closest to GHP2018, QSF to QSF ofc, which both require the KDF to be a 'good' random oracle (and QSF in the standard model requires a 'secure' PRF) which seems to basically mean 'indiff. from an RO' now, even if it's assumed/understated? Chempat doesn't have a proof but maybe we can deduce requirements from GHP2018?

> oracle.

Reviewer 4: I think the proposed change is an improvement. QROM is clearly the strongest requirement on a hash function, and for the PQ KEMs I checked, they all require one anyway.

Let me entertain a brief tangent. If I could go back a decade or so, I'd propose the CFRG to write a document with the title "What is a good hash?". It'd explain ROM, PRF, and all subtle academic differences in security properties. But in the end it'd simply say: length-prefixed-SHA2 & SHA-3 good; MD5, SHA-1 bad.

> oracle.

Author: I think it will actually be cleanest just to delete this text for now, and punt on it like we have for all of the other security descriptions.  I will also add some text at the top of this section clarifying that we're just defining functional definitions here, and the security requirements are below.

> 

Reviewer: All mention of 'KDF' and 'XOF' has been removed and 'hash functions' has been added, these things are not the same

    return DeriveKeyPair(seed)

def DeriveKeyPair(seed):
    seed_full = ExpandHash(seed)
    (seed_T, seed_PQ) = split(Group_T.Nseed, KEM_PQ.Nseed, seed)

    dk_T = Group_T.RandomScalar(seed_T))
    ek_T = Group_T.Exp(Group_T.g, dk_T)
    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)

    ek_H = concat(ek_T, ek_PQ)
    dk_H = concat(dk_T, dk_PQ)
    return (ek_H, dk_H)

def Encaps(ek):
    (ek_T, ek_PQ) = split(Group_T.Nek, KEM_PQ.Nek, ek)

    sk_E = Group_T.RandomScalar(random(GroupT.nseed))
    ct_T = Group_T.Exp(GroupT.g, sk_E)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)

    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (dk_T, dk_PQ) = split(Group_T.Ndk, KEM_PQ.Ndk, dk)
    (ct_T, ct_PQ) = split(Group_T.Nct, KEM_PQ.Nct, ct)

    ek_T = Group_T.ToEncaps(dk_T)
    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)

    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
    return ss_H
~~~

# Security Considerations

> ## Security Properties
> 
> ### INDistinguishability against Chosen-Ciphertext Attacks (IND-CCA)
> 
> ### Ciphertext Second Preimage Resistance (C2PR)
> 
> ### Binding Properties (X-BIND-P-Q)
> 
> ### Survival if One KEM Fails
> 
> ## Security of the Combiners
> 
> ### Everything {#everything-sec}
> 
> ### OnlyTraditional {#only-traditional-sec}
> 
> ### OnlySharedSecrets {#only-shared-secrets-sec}
> 
> # Security Considerations (Original)
> 

Reviewer: Removed for now but tracked in https://github.com/cfrg/draft-irtf-cfrg-hybrid-kems/issues/24

```suggestion
```

  element `p` and an integer `x`.
    * The integers `x` are called "scalars" to distinguish them from group
      elements.
    * `Exp` must respect multiplication in its scalar argument `x`, so that
      `Exp(Exp(p, x), y) = Exp(p, x * y)`.
- `RandomScalar(seed) -> k`: Produce a uniform pseudo-random scalar from the
  byte string `seed`.
- `ElementToSharedSecret(P) -> ss`: Extract a shared secret from an element of
  the group (e.g., by taking the X coordinate of an ellpitic curve point).

We assume that scalars and group elements are represented by byte strings with
fixed lengths:

- `Nseed`: The length in bytes of a seed (input to RandomScalar)
- `Nscalar`: The length in bytes of a scalar
- `Nelem`: The length in bytes of a serialized group element
- `Nss`: The length in bytes of a shared secret produced by
  ElementToSharedSecret

> It is possible to satisfy the KEM API using a nominal group, by generating an
> ephemeral element that serves as the ciphertext.  Despite satisfying the KEM
> API, however, this KEM would not be secure, so we have preferred to address
> hybrid KEMs that use elliptic curves directly.

> # Hybrid KEM Constructions {#constructions}

Reviewer: ```suggestion
# Hybrid KEM Generic Constructions {#constructions}
```

> In this section, we define three constructions for hybrid KEMs:

Reviewer: ```suggestion
In this section, we define three generic constructions for hybrid KEMs:
```

    (ct_T, ct_PQ) = split(KEM_T.Nct, KEM_PQ.Nct, ct)
    ss_T = KEM_T.Decap(dk_T, ct_T)
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))
    return ss_H
~~~

### Optimization for Encapsulation Key Reuse

The PreHashedKeys hybrid KEM is a performance optimization of the HashEverything
KEM, optimized for the case where encapsulation keys are large and frequently
reused. In such cases, hashing the entire encapsulation key is expensive, and
the same value is hashed repeatedly.  The PreHashedKeys KEM thus computes an
intermediate hash of the hybrid encapsulation key, so that the hash value can be
computed once and used across many encapsulation or decapsulation operations.

The PreHashedKeys KEM is identical to the HashEverything KEM except for the
shared secret computation.  One additional hash function is required:

* `KeyHash`: A hash function mapping byte strings of length `KEM_T.Nek +
  KEM_PQ.Nek` to byte strings of length `KEM_H.Nss` (`CombineHash.Nh ==
  KEM_H.Nss`)

The shared secret computation in `Encaps` and `Decaps` is then modified as
follows:

~~~
ekh = KeyHash(concat(ek_T, ek_PQ))

> ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))

Reviewer 3: Agreed.

~~~

> ### Instantiation with a Nominal Group

Reviewer: Let's remove this for now, tracked with #23

functions:

- `Exp(p, x) -> q`: An algorithm that produces an element `q` of `G` from an
  element `p` and an integer `x`.
    * The integers `x` are called "scalars" to distinguish them from group
      elements.
    * `Exp` must respect multiplication in its scalar argument `x`, so that
      `Exp(Exp(p, x), y) = Exp(p, x * y)`.
- `RandomScalar(seed) -> k`: Produce a uniform pseudo-random scalar from the
  byte string `seed`.
- `ElementToSharedSecret(P) -> ss`: Extract a shared secret from an element of
  the group (e.g., by taking the X coordinate of an ellpitic curve point).

We assume that scalars and group elements are represented by byte strings with
fixed lengths:

- `Nseed`: The length in bytes of a seed (input to RandomScalar)
- `Nscalar`: The length in bytes of a scalar
- `Nelem`: The length in bytes of a serialized group element
- `Nss`: The length in bytes of a shared secret produced by
  ElementToSharedSecret

> It is possible to satisfy the KEM API using a nominal group, by generating an
> ephemeral element that serves as the ciphertext.  Despite satisfying the KEM
> API, however, this KEM would not be secure, so we have preferred to address

> > hybrid KEMs that use elliptic curves directly.

Reviewer: Do we need to even say this?

## Key Derivation Functions {#kdfs}

A Key Derivation Function (KDF) is a function that a function that produces
keying material based on an input secret and other information.

> While KDFs in the literature can typically consume and produce byte strings of
> arbitrary length, the KDFs used in this document have a simpler form, with a fixed
> output lengths:
> 
> - `Nin` - The length in bytes of an input to this KDF.
> - `Nout` - The length in bytes of an output from this KDF.
> - `Derive(input) -> output`: Produce a byte string of length `Nout` from an input
>   byte string.
> 
> For simplicity, we will write invocations of a KDF without `Derive` being
> explicit.  Invoking a KDF `Foo` on input `input` will be written as `Foo(input)`
> instead of `Foo.Derive(input)`.

Reviewer: ```suggestion
While KDFs in the literature can typically consume and produce byte strings of
arbitrary length, the KDFs used in this document have a simpler form, with a fixed
output lengths:

- `Nin` - The length in bytes of an input to this KDF.
- `Nout` - The length in bytes of an output from this KDF.
- `KDF(input) -> output`: Produce a byte string of length `Nout` from an input
  byte string.

The fixed sizes are for both security and simplicity. 

For instances of the `Extract()`/`Expand()` KDF paradigm such as `HKDF`, we fix the salt s and sizes to fit this form.
```

The HashEverything hybrid KEM (`KEM_H` below) depends on the following
constituent components:

* `KEM_T`: A traditional KEM
* `KEM_PQ`: A post-quantum KEM
* `Expand`: A KDF producing byte strings of length `KEM_T.Nseed + KEM_PQ.Nseed`
  (`Expand.Nout == KEM_T.Nseed + KEM_PQ.Nseed`)
* `Combine`: A KDF producing byte strings of length `KEM_H.Nss` (`Combine.Nout
  == KEM_H.Nss`)
* `Label` - A byte string used to label the specific combination of the above
  constituents being used.

We presume the KEMs and KDFs meet the interfaces described in
{{cryptographic-deps}}.

The constants associated with the hybrid KEM are mostly derived from the
concatenation of keys and ciphertexts:

~~~
Nek = KEM_T.Nek + KEM_PQ.Nek
Ndk = KEM_T.Ndk + KEM_PQ.Ndk
Nct = KEM_T.Nct + KEM_PQ.Nct
~~~

The `Nseed` and `Nss` constants should reflect the overall security level of the
combined KEM, with the following recommended values:

~~~
Nseed = max(KEM_T.Nseed, KEM_PQ.Nseed)
Nss = min(KEM_T.Nss, KEM_PQ.Nss)
~~~

Given these constituent parts, the HashEverything hybrid KEM is defined as
follows:

~~~
def GenerateKeyPair():
    seed = random(Nseed)
    return DeriveKeyPair(seed)

def DeriveKeyPair(seed):

>     seed_full = Expand(seed)

Reviewer: Not defined

  == KEM_H.Nss`)

The `GenerateKeyPair` and `DeriveKeyPair` algorithms for PreHashedKeys are
identical to those of the HashEverything KEM.  The `Encaps` and `Decaps` method
use a modified shared secret computation:

~~~
def Encaps(ek):
    (ek_T, ek_PQ) = split(KEM_T.Nek, KEM_PQ.Nek, ek)
    (ss_T, ct_T) = KEM_T.Encap(pk_T)
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(pk_PQ)

    ekh = KeyHash(concat(ek_T, ek_PQ))
    ss_H = Combine(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))

    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (dk_T, dk_PQ) = split(KEM_T.Ndk, KEM_PQ.Ndk, dk)
    ek_T = KEM_T.ToEncaps(dk_T)
    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)

    (ct_T, ct_PQ) = split(KEM_T.Nct, KEM_PQ.Nct, ct)
    ss_T = KEM_T.Decap(dk_T, ct_T)
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ekh = KeyHash(concat(ek_T, ek_PQ))
    ss_H = Combine(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))
    return ss_H
~~~

## HashTraditionalOnly

The HashTraditionalOnly hybrid KEM (`KEM_H` below) depends on the following
constituent components:

* `Group_T`: A nominal group
* `KEM_PQ`: A post-quantum KEM
* `Expand`: A KDF producing byte strings of length `Group_T.Nseed +

>   KEM_PQ.Nseed` (`Expand.Nout == Group_T.Nseed + KEM_PQ.Nseed`)

Reviewer: QSF specifically uses an XOF

  element `p` and an integer `x`.
    * The integers `x` are called "scalars" to distinguish them from group
      elements.
    * `Exp` must respect multiplication in its scalar argument `x`, so that
      `Exp(Exp(p, x), y) = Exp(p, x * y)`.
- `RandomScalar(seed) -> k`: Produce a uniform pseudo-random scalar from the
  byte string `seed`.
- `ElementToSharedSecret(P) -> ss`: Extract a shared secret from an element of
  the group (e.g., by taking the X coordinate of an ellpitic curve point).

We assume that scalars and group elements are represented by byte strings with
fixed lengths:

- `Nseed`: The length in bytes of a seed (input to RandomScalar)
- `Nscalar`: The length in bytes of a scalar
- `Nelem`: The length in bytes of a serialized group element
- `Nss`: The length in bytes of a shared secret produced by
  ElementToSharedSecret

> It is possible to satisfy the KEM API using a nominal group, by generating an
> ephemeral element that serves as the ciphertext.  Despite satisfying the KEM
> API, however, this KEM would not be secure, so we have preferred to address
> hybrid KEMs that use elliptic curves directly.

## Key Derivation Functions {#kdfs}

> A Key Derivation Function (KDF) is a function that a function that produces
> keying material based on an input secret and other information.
> 
> While KDFs in the literature can typically consume and produce byte strings of
> arbitrary length, the KDFs used in this document have a simpler form, with a fixed
> output lengths:
> 
> - `Nin` - The length in bytes of an input to this KDF.
> - `Nout` - The length in bytes of an output from this KDF.
> - `KDF(input) -> output`: Produce a byte string of length `Nout` from an input
>   byte string.
> 
> The fixed sizes are for both security and simplicity. 
> 
> For instances of the `Extract()`/`Expand()` KDF paradigm such as `HKDF`, we fix the salt s and sizes to fit this form.
> 
> The security requirements for KDFs used with the schemes in this document are
> laid out in {{security-requirements}}.
> 
> <!-- We must include XOFs because of X-Wing at the least but also because different KEMs/ groups may need different input seed sizes out the back of the function, and pure hash functions are not well suited for this; HKDF allows this via Expand(..., length) but X-Wing does not use HKDF -->

Reviewer: ```suggestion
A Key Derivation Function (KDF) is a function that a function that produces
keying material based on an input secret and other information.

While KDFs in the literature can typically consume and produce byte strings of
arbitrary length, the KDFs used in this document have a simpler form, with a fixed
output lengths:

- `Nin` - The length in bytes of an input to this KDF.
- `Nout` - The length in bytes of an output from this KDF.
- `KDF(input) -> output`: Produce a byte string of length `Nout` from an input
  byte string.

The fixed sizes are for both security and simplicity. 

For instances of the `Extract()`/`Expand()` KDF paradigm such as `HKDF`, we fix
the salt and sizes to fit this form.

The security requirements for KDFs used with the schemes in this document are
laid out in {{security-requirements}}.

<!-- We must include XOFs because of X-Wing at the least but also because
different KEMs/ groups may need different input seed sizes out the back of
the function, and pure hash functions are not well suited for this; HKDF
allows this via Expand(..., length) but X-Wing does not use HKDF -->
```

Reviewer: Line wrap

             |                              |
             V                              V
          +-----+                        +-----+
       +->| Exp |                        | Exp |<-+
       |  +-----+                        +-----+  |
       |     |                              |     |
       |     |                              |     |
       |     V                              V     |
       |    pkA                            pkB    |
       |     |                              |     |
 skA --+     +-------------.  .-------------+     +-- skB
       |                    \/                    |
       |                    /\                    |
       |     +-------------'  '-------------+     |
       |     |                              |     |
       |     V                              V     |
       |  +-----+                        +-----+  |
       +->| Exp |                        | Exp |<-+
          +-----+                        +-----+
             |                              |
             |                              |
             V                              V
           pkAB ========================= pkBA
~~~

> Nominal groups are an abstraction that represents the core mechanics of
> Diffie-Hellman key exchange, including both finite field and elliptic-curve
> Diffie-Hellman {{ABH+21}}.  A nominal group comprises a set `G` together with a
> distinguished basis element `g`, an "exponentiation" map, and some auxiliary
> functions:

Reviewer: This is incorrect, nominal groups do not align with finite fields especially regarding Diffie-Hellman - finite fields have other algebraic structure that ofc make them weaker to instantiate DH over than groups.

```suggestion
Nominal groups are an abstract model of elliptic curve groups, over which we
instantiate Diffie-Hellman key agreement {{ABH+21}}.  A nominal group comprises
a set `G` together with a distinguished basis element `g`, an "exponentiation"
map, and some auxiliary functions:
```

  of traditional and PQ KEMs, with minimal security assumptions on the
  constituent KEMs
* PreHashedKeys - A performance optimization of HashEverything for the case
  where encapsulation keys are large and frequently reused
* HashTraditionalOnly - An optimized generic construction for the case where the traditional
  component is a nominal group and the PQ component has strong binding
  properties

In this section, we define several generic constructions for hybrid KEMs. These
constructions share a common overall structure, differing mainly in how they
compute the final shared secret and the security requirements of their components.

## HashEverything

The HashEverything hybrid KEM (`KEM_H` below) depends on the following
constituent components:

* `KEM_T`: A traditional KEM
* `KEM_PQ`: A post-quantum KEM
* `Expand`: A KDF producing byte strings of length `KEM_T.Nseed + KEM_PQ.Nseed`
  (`Expand.Nout == KEM_T.Nseed + KEM_PQ.Nseed`)
* `Combine`: A KDF producing byte strings of length `KEM_H.Nss` (`Combine.Nout
  == KEM_H.Nss`)
* `Label` - A byte string used to label the specific combination of the above
  constituents being used.

> We presume the KEMs and KDFs meet the interfaces described in
> {{cryptographic-deps}}.

Reviewer: ```suggestion
The KEMs, groups, KDFs, and XOFs MSUT meet the security requirements in {{#security-requirements}}.
```

The constants associated with the hybrid KEM are mostly derived from the
concatenation of keys and ciphertexts:

~~~
Nek = KEM_T.Nek + KEM_PQ.Nek
Ndk = KEM_T.Ndk + KEM_PQ.Ndk
Nct = KEM_T.Nct + KEM_PQ.Nct
~~~

The `Nseed` and `Nss` constants should reflect the overall security level of the
combined KEM, with the following recommended values:

~~~
Nseed = max(KEM_T.Nseed, KEM_PQ.Nseed)
Nss = min(KEM_T.Nss, KEM_PQ.Nss)

> ~~~

Reviewer: @bwesterb @br-hale @pag-crypto 

I feel like there's a little bit more to this than mix/max'ing but I'm not sure

    (ss_PQ, ct_PQ) = KEM_PQ.Encap(pk_PQ)
    ss_H = Combine(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))
    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (dk_T, dk_PQ) = split(KEM_T.Ndk, KEM_PQ.Ndk, dk)
    ek_T = KEM_T.ToEncaps(dk_T)
    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)

    (ct_T, ct_PQ) = split(KEM_T.Nct, KEM_PQ.Nct, ct)
    ss_T = KEM_T.Decap(dk_T, ct_T)
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ss_H = Combine(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))
    return ss_H
~~~

## PreHashedKeys {#prehash}

The PreHashedKeys hybrid KEM is a performance optimization of the HashEverything
KEM, optimized for the case where encapsulation keys are large and frequently
reused. In such cases, hashing the entire encapsulation key is expensive, and
the same value is hashed repeatedly.  The PreHashedKeys KEM thus computes an
intermediate hash of the hybrid encapsulation key, so that the hash value can be
computed once and used across many encapsulation or decapsulation operations.

The PreHashedKeys KEM is identical to the HashEverything KEM except for the
shared secret computation.  One additional KDF is required:

> * `KeyHash`: A KDF producing byte strings of length `KEM_H.Nss` (`KeyHash.Nout

Reviewer: `KeyHash` does not have to have the same requirements of a KDF, it probably only needs collision resistance (but I'm not sure because we don't have a proof! @pag-crypto )

  == KEM_H.Nss`)

The `GenerateKeyPair` and `DeriveKeyPair` algorithms for PreHashedKeys are
identical to those of the HashEverything KEM.  The `Encaps` and `Decaps` method
use a modified shared secret computation:

~~~
def Encaps(ek):
    (ek_T, ek_PQ) = split(KEM_T.Nek, KEM_PQ.Nek, ek)
    (ss_T, ct_T) = KEM_T.Encap(pk_T)
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(pk_PQ)

    ekh = KeyHash(concat(ek_T, ek_PQ))

>     ss_H = Combine(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))

Reviewer: ```suggestion
    ss_H = KDF(concat(ss_PQ, ss_T, ct_PQ, ct_T, ekh, label))
```

follows:

~~~
def GenerateKeyPair():
    seed = random(Nseed)
    return DeriveKeyPair(seed)

def DeriveKeyPair(seed):
    seed_full = Expand(seed)
    (seed_T, seed_PQ) = split(Group_T.Nseed, KEM_PQ.Nseed, seed)

    dk_T = Group_T.RandomScalar(seed_T))
    ek_T = Group_T.Exp(Group_T.g, dk_T)
    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)

    ek_H = concat(ek_T, ek_PQ)
    dk_H = concat(dk_T, dk_PQ)
    return (ek_H, dk_H)

def Encaps(ek):
    (ek_T, ek_PQ) = split(Group_T.Nek, KEM_PQ.Nek, ek)

    sk_E = Group_T.RandomScalar(random(GroupT.nseed))
    ct_T = Group_T.Exp(GroupT.g, sk_E)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)

    ss_H = Combine(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (dk_T, dk_PQ) = split(Group_T.Ndk, KEM_PQ.Ndk, dk)
    (ct_T, ct_PQ) = split(Group_T.Nct, KEM_PQ.Nct, ct)

    ek_T = Group_T.ToEncaps(dk_T)
    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)

    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

>     ss_H = Combine(concat(ss_PQ, ss_T, ct_T, ek_T, Label))

Reviewer: ```suggestion
    ss_H = KDF(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
```

components from a seed that is expanded to a output length, of which a subset
is passed to the component key generation algorithms.

# Hybrid KEM Schemes {#schemes}

In this section, we define three schemes for building for hybrid KEMs:

> * HashEverything - A generic construction that is suitable for use with any choice
>   of traditional and PQ KEMs, with minimal security assumptions on the
>   constituent KEMs
> * PreHashedKeys - A performance optimization of HashEverything for the case
>   where encapsulation keys are large and frequently reused
> * HashTraditionalOnly - An optimized generic construction for the case where the traditional
>   component is a nominal group and the PQ component has strong binding
>   properties

Reviewer: Concrete suggestions:

GHP - named after GHP2018 which is the foundation of all other combiners basically
PRE - prehash
QSF - same as QSF in X-Wing

Short, mostly meaningful, done.

```suggestion
* GHP - A generic construction that is suitable for use with any choice
  of traditional and PQ KEMs, with minimal security assumptions on the
  constituent KEMs
* PRE - A performance optimization of HashEverything for the case
  where encapsulation keys are large and frequently reused
* QSF - An optimized generic construction for the case where the traditional
  component is a nominal group and the PQ component has strong binding
  properties
```

@bifurcation @br-hale @bwesterb @pag-crypto

In this section, we define several generic constructions for hybrid KEMs. These
constructions share a common overall structure, differing mainly in how they
compute the final shared secret and the security requirements of their components.

## HashEverything

The HashEverything hybrid KEM (`KEM_H` below) depends on the following
constituent components:

* `KEM_T`: A traditional KEM
* `KEM_PQ`: A post-quantum KEM
* `XOF`: An XOF producing byte strings of length `KEM_T.Nseed + KEM_PQ.Nseed`
  (`XOF.Nout == KEM_T.Nseed + KEM_PQ.Nseed`)
* `KDF`: A KDF producing byte strings of length `KEM_H.Nss` (`KDF.Nout
  == KEM_H.Nss`)
* `Label` - A byte string used to label the specific combination of the above
  constituents being used.

> The KEMs, groups, KDFs, and XOFs MSUT meet the security requirements in {{#security-requirements}}.

Reviewer: ```suggestion
The KEMs, groups, KDFs, and XOFs MUST meet the security requirements in {{#security-requirements}}.
```

    return (ek_H, dk_H)

def Encaps(ek):
    (ek_T, ek_PQ) = split(Group_T.Nek, KEM_PQ.Nek, ek)

    sk_E = Group_T.RandomScalar(random(GroupT.nseed))
    ct_T = Group_T.Exp(GroupT.g, sk_E)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)

    ss_H = KDF(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
    ct_H = concat(ct_T, ct_PQ)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (dk_T, dk_PQ) = split(Group_T.Ndk, KEM_PQ.Ndk, dk)
    (ct_T, ct_PQ) = split(Group_T.Nct, KEM_PQ.Nct, ct)

    ek_T = Group_T.ToEncaps(dk_T)
    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)

    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)

    ss_H = KDF(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
    return ss_H
~~~

# Security Considerations {#security}

Hybrid KEM constructions aim to provide security by combining two or more
schemes so that security is preserved if all but one schemes are replaced by
an arbitrarily bad scheme. Informally, these hybrid KEMs are secure if the `KDF`
is secure, and either the elliptic curve is secure, or the post-quantum KEM is
secure: this is the 'hybrid' property.

> ## Security Properties

Reviewer: ```suggestion
## Security Properties {#security-properties}
```

> ### IND-CCA Security

Reviewer: ```suggestion
### IND-CCA Security {#ind-cca}
```

ciphertext, even if the two candidate messages are chosen by the adversary
himself.

> ### Ciphertext Second Preimage Resistant (C2SPI) Security
> 
> Also known in the literature as ciphertext collision resistance (CCR).

Reviewer: ```suggestion
### Ciphertext Second Preimage Resistant (C2PRI) Security

Also known in the literature as ciphertext collision resistance (CCR).
```

{{XWING}}. The same notion has also been described as chosen ciphertext
resistance elsewhere {{CDM23}}.

## Binding Properties

It is often useful for a KEM to have certain "binding properties", by which
certain parameters determine certain others {{CDM23}}.  These properties are
referred to with labels of the form X-BIND-P-Q.  For example, LEAK-BIND-K-PK
means that for a given shared secret (K), there is a unique encapsulation key
(PK) that could have produced it, even if all of the secrets involved are given
to the adversary after the encapsulation operation is completed (LEAK).

The property LEAK-BIND-K,PK-CT is related to the C2PRI property discussed above.
Related to the ciphertext collision-freeness of the underlying PKE scheme of a
FO-transform KEM. Also called ciphertext collision resistance.

[[ TODO: Discuss other salient binding properties. ]]

## Security Requirements for Constituent Components {#security-requirements}

> TODO: We need to provide more thorough description, and verify that
> these requirements align with the requirements of the security proofs in the
> literature, especially {{GHP2018}} and {{XWING}}.

> ### Security Requirements for KEMs

Reviewer: ```suggestion
### Security Requirements for KEMs {#security-kems}
```

Component KEMs MUST be IND-CCA-secure {{GHP2018}} {{XWING}}.

For instances of QSF, the component KEM MUST also be ciphertext second
preimage resistant (C2PRI) {{XWING}}: this allows the component KEM
encapsulation key and ciphertext to be left out from the KDF input. 

### Security Requirements for Groups

The groups MUST be modelable as nominal groups in which the strong
Diffie-Hellman problem holds {{ABH+21}} {{XWING}}.

The Montgomery curves Curve25519 and Curve448 have been shown to be
modelable as nominal groups in {{ABH+21}} as well as showing the 
`X25519()` and `X448()` functions respectively pertain to the nominal
group `exp(X, y)` function, specifically clamping secret keys when
they are generated, instead of clamping secret keys together with
exponentiation.

<!-- The short Weierstrass NIST curves have also been shown to be
modelable as nominal groups but I can't find the reference --> 

> ### Security Requirements for KDFs

Reviewer: ```suggestion
### Security Requirements for KDFs {#security-kdfs}
```

still remains the best way to rule out structural attacks.

Sponge-based constructions such as SHA-3 have been shown to be
indifferentiable against classical {{BDP+08 https://www.iacr.org/archive/eurocrypt2008/49650180/49650180.pdf}}
as well as quantum adversaries 
{{ACM+25 https://eprint.iacr.org/2025/731.pdf}}. 

HKDF has been shown to be indifferentiable from a random oracle under
specific constraints {{LBB20 https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8806752}}: 

- that HMAC is indifferentiable from a random oracle,
which for HMAC-SHA-256 has been shown in {{}} {{DRS+13 https://eprint.iacr.org/2013/382.pdf}}, assuming the
compression function underlying SHA-256 is a random oracle,
which it is indifferentiably when used prefix-free.

- the values of `HKDF`'s `IKM` input do not collide with
values of `info` `||` `0x01`. This MUST be enforced by the
concrete instantiations that use `HKDF` as its KDF. 

The choice of the KDF security level SHOULD be made based on the
security level provided by the constituent KEMs. The KDF SHOULD
at least have the security level of the strongest constituent KEM.

> ### Security Requirements for XOFs

Reviewer: ```suggestion
### Security Requirements for XOFs {#security-xofs}
```

XOFs accept arbitrary bitstrings as input, and produce
a caller-chosen-length prefix of an infinite bitstream
deterministically defined by the input.

> <!-- requirements: secure PRF, bc key material? -->

Reviewer: @br-hale @pag-crypto @bwesterb

{{XWING}}. The same notion has also been described as chosen ciphertext
resistance elsewhere {{CDM23}}.

> ## Binding Properties

Reviewer: ```suggestion
## Binding Properties {#binding-properties}
```

{{XWING}}. The same notion has also been described as chosen ciphertext
resistance elsewhere {{CDM23}}.

## Binding Properties

It is often useful for a KEM to have certain "binding properties", by which
certain parameters determine certain others {{CDM23}}.  These properties are
referred to with labels of the form X-BIND-P-Q.  For example, LEAK-BIND-K-PK
means that for a given shared secret (K), there is a unique encapsulation key
(PK) that could have produced it, even if all of the secrets involved are given
to the adversary after the encapsulation operation is completed (LEAK).

The property LEAK-BIND-K,PK-CT is related to the C2PRI property discussed above.
Related to the ciphertext collision-freeness of the underlying PKE scheme of a
FO-transform KEM. Also called ciphertext collision resistance.

[[ TODO: Discuss other salient binding properties. ]]

## Security Requirements for Constituent Components {#security-requirements}

> > TODO: The text in this section is provisional, intended to give a general idea of
> > direction.  We need to provide more thorough description, and verify that
> > these requirements align with the requirements of the security proofs in the
> > literature, especially {{GHP2018}} and {{XWING}}.
> 
> KEMs must be IND-CCA, except for in HashTraditionalOnly, where they must provide
> certain binding properties as specified in {{XWING}}.
> 
> Nominal groups must be ones in which appropriate Diffie-Hellman hardness results
> apply.
> 
> KDFs must be random oracles in the ROM and QROM.  (This seems to be the
> assumption made in both {{GHP2018}} and {{XWING}}.)  The choice of the KDF for
> HPKE SHOULD be made based on the security level provided by the constituent
> KEMs. The KDF SHOULD at least have the security level of the strongest
> constituent KEM.

Reviewer: ```suggestion
> TODO: We need to provide more thorough description, and verify that
> these requirements align with the requirements of the security proofs in the
> literature, especially {{GHP2018}} and {{XWING}}.

### Security Requirements for KEMs

Component KEMs MUST be IND-CCA-secure {{GHP2018}} {{XWING}}.

For instances of QSF, the component KEM MUST also be ciphertext second
preimage resistant (C2PRI) {{XWING}}: this allows the component KEM
encapsulation key and ciphertext to be left out from the KDF input. 

### Security Requirements for Groups

The groups MUST be modelable as nominal groups in which the strong
Diffie-Hellman problem holds {{ABH+21}} {{XWING}}.

The Montgomery curves Curve25519 and Curve448 have been shown to be
modelable as nominal groups in {{ABH+21}} as well as showing the 
`X25519()` and `X448()` functions respectively pertain to the nominal
group `exp(X, y)` function, specifically clamping secret keys when
they are generated, instead of clamping secret keys together with
exponentiation.

<!-- The short Weierstrass NIST curves have also been shown to be
modelable as nominal groups but I can't find the reference --> 

### Security Requirements for KDFs

KDFs MUST be secure pseudorandom functions (PRFs) when keyed with
the shared secret output from the post-quantum IND-CCA-secure
KEM component algorithm in QSF {{XWING}} or any of the component
IND-CCA-secure KEMs when used in KitchenSink {{GHP2018}} or
PreHash.

KDFs must be secure instances of random oracles in the ROM and
QROM {{GHP2018}} {{XWING}}. Proofs of indifferentiability from
random oracles {{MRH03 https://eprint.iacr.org/2003/161.pdf}}
give good confidence here, as any function proven indifferentiable
from a random oracle is resistant against collision, first, and
second preimage attacks {{need a good cite here}}. An indifferentiability
bound guarantees security against specific attacks. Although
indifferentiability does not capture all properties of a random
oracle {{RSS11 https://eprint.iacr.org/2011/339.pdf}}, indifferentiability
still remains the best way to rule out structural attacks.

Sponge-based constructions such as SHA-3 have been shown to be
indifferentiable against classical {{BDP+08 https://www.iacr.org/archive/eurocrypt2008/49650180/49650180.pdf}}
as well as quantum adversaries 
{{ACM+25 https://eprint.iacr.org/2025/731.pdf}}. 

HKDF has been shown to be indifferentiable from a random oracle under
specific constraints {{LBB20 https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8806752}}: 

- that HMAC is indifferentiable from a random oracle,
which for HMAC-SHA-256 has been shown in {{}} {{DRS+13 https://eprint.iacr.org/2013/382.pdf}}, assuming the
compression function underlying SHA-256 is a random oracle,
which it is indifferentiably when used prefix-free.

- the values of `HKDF`'s `IKM` input do not collide with
values of `info` `||` `0x01`. This MUST be enforced by the
concrete instantiations that use `HKDF` as its KDF. 

The choice of the KDF security level SHOULD be made based on the
security level provided by the constituent KEMs. The KDF SHOULD
at least have the security level of the strongest constituent KEM.

### Security Requirements for XOFs

XOFs accept arbitrary bitstrings as input, and produce
a caller-chosen-length prefix of an infinite bitstream
deterministically defined by the input.

<!-- requirements: secure PRF, bc key material? -->

```
